### PR TITLE
Type definitions for parity-pmr

### PR DESCRIPTION
**Pull request by leovujanic**
Hi @jvirtanen! 

First of all, thank you for amazing work on Parity and everything related to it. 

Here I have added type definitions so that this module can be used in TypeScript. Types are written following [Definitely Typed](http://definitelytyped.org/) guidelines and they are tested with their linter.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    orderNumber?: string;
+    side?: string;
+    instrument?: string;
+    quantity?: number;
+    price?: number;
+    canceledQuantity?: number;
+    matchNumber?: number;
+    restingOrderNumber?: number;
+    incomingOrderNumber?: number;
+}
```
Comment (jvirtanen) [inline index.d.ts]: Would it make sense to declare `PMRMessage` as a union of `PMRVersion`, `PMROrderEntered`, `PMROrderAdded`, `PMROrderCanceled`, and `PMRTrade`?

**Comment by jvirtanen** [comment]
Thank you, @leovujanic! I haven't done anything with TypeScript myself yet, but it's definitely on the to-do list. 👍

**Comment by leovujanic** [comment]
Hi @jvirtanen , yes TypeScript is great and I hope that feature of javascript will look just like that. :)

Regarding your question about message types, I really like your suggestion.
Here is what I can do:

```typescript
export enum PMRMessageType {
    ORDER_ADDED = 'A',
    ORDER_ENTERED = 'E',
    ORDER_CANCELED = 'X',
    VERSION = 'V',
    TRADE = 'T',
}

export enum PMRSideEnum {
    BUY = 'B',
    SELL = 'S'
}

declare interface PMRBaseMessage {
    messageType: PMRMessageType;
}

export interface PMRVersion extends PMRBaseMessage {
    version: number;
}

export interface PMROrderEntered  extends PMRBaseMessage {
    timestamp: number;
    username: string;
    orderNumber: string;
    side: PMRSideEnum;
    instrument: string;
    quantity: number;
    price: number;
}


export function format(message: PMRVersion | PMROrderEntered): Buffer;

export function parse(buffer: Buffer): PMRVersion | PMROrderEntered;
```

And usage would be like this:
```typescript
const messageToFormat: PMRVersion = {
    messageType: PMRMessageType.VERSION,
    version: 1,
};

const buffer = PMR.format(messageToFormat);

// here we don't know type
const message = PMR.parse(new Buffer("test"));

switch(message.messageType) {
    case PMRMessageType.VERSION:
        // here we need to cast to PMRVersion 
        const versionMessage : PMRVersion = <PMRVersion>message;
        break;
}
```

If you are OK with this solution I can make changes during the day and apply the same principle to PMD and POE libraries.

**Comment by jvirtanen** [comment]
I think either way would work. Which is better from a TypeScript writer's perspective?

How does it look when you use these definitions? Are they part of a namespace, similarly as the `format()` and `parse()` methods when using the JavaScript interface directly?
```javascript
const PMR = require('parity-pmr');

PMR.parse(buffer);
```
If that is the case, maybe we could drop the "PMR" prefix from the types.

**Comment by leovujanic** [comment]
I think we should go with the second solution (enums and message types) it's more strict.  The same principle will gain more benefit in POE library where are multiple message available with the same `messageType` but one for inbound and one for outbound.

We can omit PMR prefix and stick with usage like this:

```typescript
import * as PMR from "parity-pmr";

const versionMessage: PMR.Version = PMR.parse(new Buffer("test"));

const message: PMR.OrderEntered = {
    messageType: PMR.MessageType.ORDER_ADDED,
    orderNumber: 1234,
    side: PMR.SideEnum.BUY,
    ...
};
```

The only thing that I don't like is message type naming. Would it be better to define it as `VersionMessage` instead of `Version`? Just `Version` doesn't speak much for itself. 

What do you think?

**Comment by jvirtanen** [comment]
I think the latest sketch looks great! 👍 

Regarding the naming, as a precedence, names without a "Message" suffix are used in the [protocol specification](https://github.com/paritytrading/parity/blob/master/libraries/net/doc/PMR.md#version) as well as the [reference implementation](https://github.com/paritytrading/parity/blob/master/libraries/net/src/main/java/com/paritytrading/parity/net/pmr/PMRListener.java#L18). So keeping it as-is would be in line with the existing conventions.

As a little nit, I would call `PMR.SideEnum` just `PMR.Side` to be consistent with the other types. 🙂

**Comment by leovujanic** [comment]
Yes, I agree with everything. I will make changes and push later today.

**Comment by leovujanic** [comment]
I have added an interface for each message, but I was unable to use enums so I used string literals. TS expects if an enum is defined that it exists in js. This is also ok because the string value will be linted. 
The usage for this version would be like this:
```typescript
import * as PMR from "parity-pmr";

const orderEntered : PMR.OrderEntered = {
    messageType: "E" ,
    timestamp: 123456,
    username: "username",
    orderNumber: 123456,
    side: "B",
    instrument: "ETH-BTC",
    quantity: 2,
    price: 30,
};

const formattedMessage: Buffer = PMR.format(orderEntered);

const parsedMessage = PMR.parse(formattedMessage);

const oe: PMR.OrderEntered = <PMR.OrderEntered>parsedMessage;
```

I also created PR with small changes in the library so that we can use it with enums. Please, take a look at #4 . Can we go with that version?

**Pull request merged**